test/suites/byok/suite_test.go

```diff
@@ -126,3 +136,235 @@ var _ = Describe("BYOK", func() {
+)
+
+// CreateKeyVaultAndDiskEncryptionSet creates a key vault and disk encryption set for BYOK testing
+//
+// This function performs the following operations in sequence:
+//
+// 1. API Calls:
+//   - Creates an Azure Key Vault with RBAC authorization enabled
+//   - Creates a key in the Key Vault for encryption
+//   - Creates a Disk Encryption Set (DES) referencing the Key Vault key
+//
+// 2. RBAC Assignments:
+//   - Assigns Key Vault Crypto Officer and Administrator roles to the Karpenter workload identity
+//   - Assigns Key Vault Crypto Officer and Administrator roles to the test user (for key creation)
+//   - Assigns Reader role to the Karpenter workload identity for DES access (so it can create VMs)
+//   - Assigns Key Vault Crypto Service Encryption User role to the DES managed identity (so DES can access keys)
+//
+// 3. Retry Logic:
+//   - Azure clients are configured with retry logic for 403 (Forbidden) errors caused by RBAC propagation delays
+//   - Azure clients also retry on 400 (Bad Request) errors for PrincipalNotFound when identities haven't replicated
+//
+// Returns the DES resource ID for use in VM disk encryption configuration.
+func CreateKeyVaultAndDiskEncryptionSet(ctx context.Context, env *azure.Environment) string {
+	keyVaultName := fmt.Sprintf("karpentertest%d", time.Now().Unix())
+	keyName := "test-key"
+	desName := fmt.Sprintf("karpenter-test-des-%d", time.Now().Unix())
+
+	cred := env.GetDefaultCredential()
+
+	clusterIdentity := env.GetClusterIdentity(ctx)
+	clusterTenant := lo.FromPtr(clusterIdentity.TenantID)
+	Expect(clusterTenant).ToNot(BeEmpty())
+
+	karpenterIdentity := env.GetKarpenterWorkloadIdentity(ctx)
+
+	keyVault, err := createKeyVault(ctx, env, keyVaultName, clusterTenant)
+	Expect(err).ToNot(HaveOccurred())
+
+	// Get the current test user's principal ID from the defaultCredential Token
+	testUserPrincipalID := env.GetCurrentUserPrincipalID(ctx, cred)
+	Expect(testUserPrincipalID).ToNot(BeEmpty(), "test user authentication failed")
+
+	err = assignKeyVaultRBAC(ctx, env, lo.FromPtr(keyVault.ID), karpenterIdentity, testUserPrincipalID)
+	Expect(err).ToNot(HaveOccurred())
+
+	key, err := createKeyVaultKey(ctx, keyVaultName, keyName, cred)
+	Expect(err).ToNot(HaveOccurred())
+
+	des, err := createDiskEncryptionSet(ctx, env, desName, keyVault, key)
+	Expect(err).ToNot(HaveOccurred())
+
+	desIdentity := des.Identity
+	Expect(desIdentity).ToNot(BeNil())
+	Expect(desIdentity.PrincipalID).ToNot(BeNil())
+
+	err = assignDiskEncryptionSetRBAC(ctx, env, lo.FromPtr(des.ID), karpenterIdentity)
+	Expect(err).ToNot(HaveOccurred())
+
+	err = updateKeyVaultAccessForDES(ctx, env, keyVaultName, desIdentity)
+	Expect(err).ToNot(HaveOccurred())
+
+	return lo.FromPtr(des.ID)
+}
+
+// createKeyVault creates an Azure Key Vault
+func createKeyVault(ctx context.Context, env *azure.Environment, keyVaultName, clusterTenant string) (*armkeyvault.Vault, error) {
```

Comment:
Ah sorry my previous comment wasn't clear maybe.

I was just saying that _one method_ `CreateKeyVaultAndDiskEncryptionSet` doesn't fit on env.
I'm fine with `createKeyVault` and `assignKeyVaultRBAC` being exported and on env. They feel reusable for other tests than this one.

even `createDiskEncryptionSet` could probably be on env (who knows maybe we'll re-use it?)

`CreateKeyVaultAndDiskEncryptionSet` though felt more like "this is specific setup for a specific test" and so was less reusable.